engine: pass rocksdb perf stats back to go

tbg opened the pull request:
Being able to see RocksDB-level perf counters in slow read operations
can be helpful to diagnose issues such as #17229. Being able to see
I/O slowness at the RocksDB level should be helpful to diagnose disk
bottlenecking.

This commit optionally enables collection of (a currently singleton
subset) RocksDB performance counters on a per-iterator basis but
makes it straightforward to add anything from the IO or PerfContext.

Note that the RocksDB perf counters use TLS (thread-local storage).
Iterators are usually created and used from goroutines, so they are not
tied to a thread. To avoid this, the iterator holds a struct into which
it accumulates the collected counters for each operation that takes
place in C++. Compared to the overhead of crossing the cgo barrier in
the first place, the added two conditionals seem negligible (and the
branch predictor should do a good job on them).

As a first cut, it prints the number of skipped RocksDB tombstones
into the trace for scan operations.

Was was tempted to make this event log automatic upon closing the
iterator, but this would require passing a `context.Context` into
`(Iterator).Close` as well as `NewIterator` which seemed potentially
controversial.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/24543)
<!-- Reviewable:end -->

petermattis commented:
I think we definitely want to avoid bloating the `DBScanResults` struct. Returning a pointer to be freed is one option. Or that pointer could refer to a `PerfContext` stored in the iterator. 

What mechanism are you imagining will enable this? How much of the RocksDB perf context should we provide access to? I certainly found utility to looking at the perf context stuff when optimizing `MVCCScan`, but exposing all of it seems excessive.

---

Review status: 0 of 2 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24543#-:-L9QDMc31iAsG8lFp1hU:bed34c1)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
@petermattis I'm mostly interested in exposing the tombstone counts for #17229. The perf context would be enabled when a trace is present.

petermattis commented:
> @petermattis I'm mostly interested in exposing the tombstone counts for #17229. The perf context would be enabled when a trace is present.

Ok. My preference would be to only add the fields to `PerfContext` that you're interested in, along with a TODO that more fields can be added as necessary.

tschottdorf commented:
Ready for another look!

petermattis commented:
That was a lot of plumbing. I'll need to take another look at this tomorrow.

---

Review status: 0 of 42 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[c-deps/libroach/db.cc, line 99 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24543#-L9rRHoPC3VP4b0cnRdK:-L9rRHoPC3VP4b0cnRdL:bg024l9) ([raw file](https://github.com/cockroachdb/cockroach/blob/58c6450fe75ce0d754796fca2eafd78771ba6950/c-deps/libroach/db.cc#L99)):*
> ```cpp
>   }
> 
>   DBIterAggStats(iter);
> ```

Hmm, this is a bit subtle as this call is distant from the corresponding `DBIterSetupStats`. Another approach would be a small class that you drop into all of the `DBIter*` functions and perfom the `DBIterSetupStats` and `DBIterAggStats` work in the constructor and destructor. Something like:

```
struct ScopedStats {
  ScopedStats(DBIterator *iter)
    iter_(iter) {
    DBIterSetupStats(iter_); // should inline the DBIterSetupStats code here
  }
  ~ScopedStats() {
    DBIterAggStats(iter_);
  }
};
...
DBIterState DBIterSeek(DBIterator* iter, DBKey key) {
  ScopedStats stats(iter_);
  ...
}
```

---

*[c-deps/libroach/db.cc, line 404 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24543#-L9rQLEpCKGlKb-XoV64:-L9rQLEpCKGlKb-XoV65:b-gfxq0u) ([raw file](https://github.com/cockroachdb/cockroach/blob/58c6450fe75ce0d754796fca2eafd78771ba6950/c-deps/libroach/db.cc#L404)):*
> ```cpp
>   if (stats) {
>     db_iter->stats.reset(new IteratorStats);
>     *db_iter->stats = {};
> ```

Does this actually zero the stats? Usually I use something like `memset(db_iter->stats.get(), 0, sizeof(IteratorStats))`.

---

*[c-deps/libroach/db.cc, line 433 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24543#-L9rOQ8t6V_SqOM1Y5lK:-L9rOQ8t6V_SqOM1Y5lL:b-qf2onb) ([raw file](https://github.com/cockroachdb/cockroach/blob/58c6450fe75ce0d754796fca2eafd78771ba6950/c-deps/libroach/db.cc#L433)):*
> ```cpp
>   // TODO(tschottdorf): plumb this in. We should be able to figure out
>   // how many SSTables this touches, which will be interesting. I'm not
>   // sure which of the stats indicates that number, though.
> ```

Perhaps we can just count the accesses directly in the `table_filter`. 

---

*[c-deps/libroach/db.cc, line 436 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24543#-L9rQIpo7M664TsuHsD1:-L9rQIpo7M664TsuHsD2:b-fkxpqc) ([raw file](https://github.com/cockroachdb/cockroach/blob/58c6450fe75ce0d754796fca2eafd78771ba6950/c-deps/libroach/db.cc#L436)):*
> ```cpp
>   bool stats = false;
>   if (stats) {
>     db_iter->stats.reset(new IteratorStats);
> ```

You're not zeroing `db_iter->stats` here.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24543#-:-L9rP4rq7q1UREIyWvwZ:b-8bdb4u)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Looks good modulo my comments.

---

Review status: 0 of 42 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[c-deps/libroach/db.h, line 75 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24543#-L9u3YYj1_McL-R-ladD:-L9u3YYk0lV7Am0Y7-gn:bs934io) ([raw file](https://github.com/cockroachdb/cockroach/blob/58c6450fe75ce0d754796fca2eafd78771ba6950/c-deps/libroach/db.h#L75)):*
> ```objc
> 
> void DBIterSetupStats(DBIterator* iter);
> void DBIterAggStats(DBIterator* iter);
> ```

If these are only used in `db.cc`, they don't need to be declared in this header file. My `ScopedStats` would get rid of them completely, though.

---

*[c-deps/libroach/db.cc, line 64 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24543#-L9u4DAg92hIRE5qepRY:-L9u4DAg92hIRE5qepRZ:b-gegdte) ([raw file](https://github.com/cockroachdb/cockroach/blob/58c6450fe75ce0d754796fca2eafd78771ba6950/c-deps/libroach/db.cc#L64)):*
> ```cpp
> void DBIterSetupStats(DBIterator* iter) {
>   if (iter->stats != nullptr) {
>     rocksdb::get_perf_context()->Reset();
> ```

The perf-contexts are per-thread, right? Rather than calling `Reset` which needs to zero a lot of fields, it might be better to store `pctx->internal_delete_skipped_count` in a local variable (see `ScopedStats` suggestion) and then compute the delta in `DBIterAggStats`.

---

*[c-deps/libroach/db.cc, line 68 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24543#-L9u3mRE31-D9M5CIadA:-L9u3mRE31-D9M5CIadB:b3ypwy6) ([raw file](https://github.com/cockroachdb/cockroach/blob/58c6450fe75ce0d754796fca2eafd78771ba6950/c-deps/libroach/db.cc#L68)):*
> ```cpp
>     // TODO(tschottdorf): managing the perf level is a little awkward.
>     // Is this (and the sibling in DBIterAggStats) the best we can do?
>     rocksdb::SetPerfLevel(rocksdb::PerfLevel::kEnableTimeExceptForMutex);
> ```

How does this interact with multiple iterators concurrently manipulating the stats level? Do you need to have some sort of atomic count of the number of iterators that want perf stats? 

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24543#-:-L9u5DT4DIfszqva_B0J:biu3oe)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
PTAL!

---

Review status: 0 of 15 files reviewed at latest revision, 7 unresolved discussions.

---

*[c-deps/libroach/db.h, line 75 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24543#-L9u3YYj1_McL-R-ladD:-L9ytSH_824Gni5FKg3I:bf9y48d) ([raw file](https://github.com/cockroachdb/cockroach/blob/58c6450fe75ce0d754796fca2eafd78771ba6950/c-deps/libroach/db.h#L75)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

If these are only used in `db.cc`, they don't need to be declared in this header file. My `ScopedStats` would get rid of them completely, though.
</blockquote></details>

It's used in `mvcc.h` (and so now `ScopedStats` is declared in `db.h`). Can we do any better?

---

*[c-deps/libroach/db.cc, line 64 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24543#-L9u4DAg92hIRE5qepRY:-LA-CMN7A0Bn2nSZAAZ_:b-fznogt) ([raw file](https://github.com/cockroachdb/cockroach/blob/58c6450fe75ce0d754796fca2eafd78771ba6950/c-deps/libroach/db.cc#L64)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

The perf-contexts are per-thread, right? Rather than calling `Reset` which needs to zero a lot of fields, it might be better to store `pctx->internal_delete_skipped_count` in a local variable (see `ScopedStats` suggestion) and then compute the delta in `DBIterAggStats`.
</blockquote></details>

Good idea, done.

---

*[c-deps/libroach/db.cc, line 68 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24543#-L9u3mRE31-D9M5CIadA:-L9yXnH5ENmBUBf0xooK:b-pkq0w4) ([raw file](https://github.com/cockroachdb/cockroach/blob/58c6450fe75ce0d754796fca2eafd78771ba6950/c-deps/libroach/db.cc#L68)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

How does this interact with multiple iterators concurrently manipulating the stats level? Do you need to have some sort of atomic count of the number of iterators that want perf stats? 
</blockquote></details>

Maybe I'm misunderstanding how this works. The perf counters use a thread-local, whereas we pass the iterator across threads (by using it repeatedly from a goroutine). So every time we enter C++, I reset the stats (of the current thread) and save the counter before returning into Go. If we used multiple iterators at once across a C++ call (i.e. if we had a multi iterator scan operation or something like that) then yes it would give the count for all of the iterators in each of the iterator. But I don't think we do that, right? Is that what you mean?

---

*[c-deps/libroach/db.cc, line 99 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24543#-L9rRHoPC3VP4b0cnRdK:-L9yqt8p7MKKF5txwk3v:bv8x82v) ([raw file](https://github.com/cockroachdb/cockroach/blob/58c6450fe75ce0d754796fca2eafd78771ba6950/c-deps/libroach/db.cc#L99)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Hmm, this is a bit subtle as this call is distant from the corresponding `DBIterSetupStats`. Another approach would be a small class that you drop into all of the `DBIter*` functions and perfom the `DBIterSetupStats` and `DBIterAggStats` work in the constructor and destructor. Something like:

```
struct ScopedStats {
  ScopedStats(DBIterator *iter)
    iter_(iter) {
    DBIterSetupStats(iter_); // should inline the DBIterSetupStats code here
  }
  ~ScopedStats() {
    DBIterAggStats(iter_);
  }
};
...
DBIterState DBIterSeek(DBIterator* iter, DBKey key) {
  ScopedStats stats(iter_);
  ...
}
```
</blockquote></details>

Done, so much better!

---

*[c-deps/libroach/db.cc, line 404 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24543#-L9rQLEpCKGlKb-XoV64:-L9yrlx22dh-LOXwiyZ-:bdcq40a) ([raw file](https://github.com/cockroachdb/cockroach/blob/58c6450fe75ce0d754796fca2eafd78771ba6950/c-deps/libroach/db.cc#L404)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Does this actually zero the stats? Usually I use something like `memset(db_iter->stats.get(), 0, sizeof(IteratorStats))`.
</blockquote></details>

Yeah, this should do the right thing (needs C++11) : http://en.cppreference.com/w/cpp/language/zero_initialization. I can change it if you prefer memset.

---

*[c-deps/libroach/db.cc, line 433 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24543#-L9rOQ8t6V_SqOM1Y5lK:-LA-Ca_34TRtXfJ4lq7u:brrqtcf) ([raw file](https://github.com/cockroachdb/cockroach/blob/58c6450fe75ce0d754796fca2eafd78771ba6950/c-deps/libroach/db.cc#L433)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Perhaps we can just count the accesses directly in the `table_filter`. 
</blockquote></details>

Good idea, that was actually pretty straightforward.

---

*[c-deps/libroach/db.cc, line 436 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24543#-L9rQIpo7M664TsuHsD1:-L9ysr1hEgurOeSrUCAz:b8rmskf) ([raw file](https://github.com/cockroachdb/cockroach/blob/58c6450fe75ce0d754796fca2eafd78771ba6950/c-deps/libroach/db.cc#L436)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

You're not zeroing `db_iter->stats` here.
</blockquote></details>

Good catch. Is there a one-liner for this?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24543#-:-LA-Cod944IXDsRNvfir:b-83330z)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 14 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[c-deps/libroach/db.h, line 75 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24543#-L9u3YYj1_McL-R-ladD:-LA0kH9v5yxCXHXJeYva:bkjrxwz) ([raw file](https://github.com/cockroachdb/cockroach/blob/58c6450fe75ce0d754796fca2eafd78771ba6950/c-deps/libroach/db.h#L75)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

It's used in `mvcc.h` (and so now `ScopedStats` is declared in `db.h`). Can we do any better?
</blockquote></details>

Nope. My comment was incorrect. This is good. Minor style issues here. This should look like:

```
class ScopedStats {
 public:
   ScopedStats(DBIterator*);
   ~ScopedStats();

 private:
   DBIterator* const iter_;
   uint64_t internal_delete_skipped_count_base_;
}
```

`class` and `struct` are identical except for the default visibility (private in `class`, public in `struct`). Generally with the use of constructors and private fields you'll use class. The ordering of methods and member variables follows the C++ coding guidelines. All nits and not functionally important.

---

*[c-deps/libroach/db.cc, line 68 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24543#-L9u3mRE31-D9M5CIadA:-LA0kxkV-quQ21JwSE8S:b-ce2lss) ([raw file](https://github.com/cockroachdb/cockroach/blob/58c6450fe75ce0d754796fca2eafd78771ba6950/c-deps/libroach/db.cc#L68)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Maybe I'm misunderstanding how this works. The perf counters use a thread-local, whereas we pass the iterator across threads (by using it repeatedly from a goroutine). So every time we enter C++, I reset the stats (of the current thread) and save the counter before returning into Go. If we used multiple iterators at once across a C++ call (i.e. if we had a multi iterator scan operation or something like that) then yes it would give the count for all of the iterators in each of the iterator. But I don't think we do that, right? Is that what you mean?
</blockquote></details>

You're correct. I was thinking that you were enabling the perf stats across Cgo calls when I wrote that. I later realized you weren't and never removed the comment (or event remembered I wrote it).

---

*[c-deps/libroach/db.cc, line 404 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24543#-L9rQLEpCKGlKb-XoV64:-LA0lNHq2hGHayg2e6yP:b-mxfhuk) ([raw file](https://github.com/cockroachdb/cockroach/blob/58c6450fe75ce0d754796fca2eafd78771ba6950/c-deps/libroach/db.cc#L404)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Yeah, this should do the right thing (needs C++11) : http://en.cppreference.com/w/cpp/language/zero_initialization. I can change it if you prefer memset.
</blockquote></details>

This syntax is much better than `memset`. I didn't even realize this was a thing. My C++ knowledge is getting out of date.

---

*[c-deps/libroach/db.cc, line 436 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24543#-L9rQIpo7M664TsuHsD1:-LA0lqM69tTOT5gdRjFb:btabxoc) ([raw file](https://github.com/cockroachdb/cockroach/blob/58c6450fe75ce0d754796fca2eafd78771ba6950/c-deps/libroach/db.cc#L436)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Good catch. Is there a one-liner for this?
</blockquote></details>

No one-liner I know of.

---

*[c-deps/libroach/db.cc, line 67 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24543#-L9qc4_BYiX92vykea03-r2-67:-LA0lElEBEgVjBLtsGNg:b-yi366f) ([raw file](https://github.com/cockroachdb/cockroach/blob/45d4c234a5a5475e299e5f3b8816a7ea1af026cf/c-deps/libroach/db.cc#L67)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Should we try to get rid of those by running them in an "init" function of sorts? Don't see how that can work, though.
</blockquote></details>

What are you wondering if you can get rid of? The calls to `SetPerfLevel`?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24543#-:-LA0mK4d6dBUYjLR7lNr:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
TFTR!

---

Review status: 0 of 17 files reviewed at latest revision, 5 unresolved discussions.

---

*[c-deps/libroach/db.h, line 75 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24543#-L9u3YYj1_McL-R-ladD:-LA87m2O0NdVzh8dZjkB:bo3vl6l) ([raw file](https://github.com/cockroachdb/cockroach/blob/58c6450fe75ce0d754796fca2eafd78771ba6950/c-deps/libroach/db.h#L75)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Nope. My comment was incorrect. This is good. Minor style issues here. This should look like:

```
class ScopedStats {
 public:
   ScopedStats(DBIterator*);
   ~ScopedStats();

 private:
   DBIterator* const iter_;
   uint64_t internal_delete_skipped_count_base_;
}
```

`class` and `struct` are identical except for the default visibility (private in `class`, public in `struct`). Generally with the use of constructors and private fields you'll use class. The ordering of methods and member variables follows the C++ coding guidelines. All nits and not functionally important.
</blockquote></details>

Thanks, done!

---

*[c-deps/libroach/db.cc, line 67 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24543#-L9qc4_BYiX92vykea03-r2-67:-LA87xRp69q0F1NNad3y:b-mkv3if) ([raw file](https://github.com/cockroachdb/cockroach/blob/45d4c234a5a5475e299e5f3b8816a7ea1af026cf/c-deps/libroach/db.cc#L67)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

What are you wondering if you can get rid of? The calls to `SetPerfLevel`?
</blockquote></details>

Yeah, but not sure what else I was thinking. This looks fine. I'm just worried that changing the perf level so often is not cheap in itself. It's just setting a TLS though, so that's basically free!

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24543#-:-LA8HDQw3U4vL9JNOALJ:b-814g8k)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> (again)

---

Review status: 0 of 18 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24543#-:-LAADVLK4D3U7RXeL2ps:b-ptz332)*
<!-- Sent from Reviewable.io -->

tbg committed:
engine: pass rocksdb perf stats back to go

Being able to see RocksDB-level perf counters in slow read operations
can be helpful to diagnose issues such as #17229. Being able to see
I/O slowness at the RocksDB level should be helpful to diagnose disk
bottlenecking.

This commit optionally enables collection of (a currently singleton
subset) RocksDB performance counters on a per-iterator basis but
makes it straightforward to add anything from the IO or PerfContext.

Note that the RocksDB perf counters use TLS (thread-local storage).
Iterators are usually created and used from goroutines, so they are not
tied to a thread. To avoid this, the iterator holds a struct into which
it accumulates the collected counters for each operation that takes
place in C++. Compared to the overhead of crossing the cgo barrier in
the first place, the added two conditionals seem negligible (and the
branch predictor should do a good job on them).

As a first cut, it prints the number of skipped RocksDB tombstones
into the trace for scan operations.

Was was tempted to make this event log automatic upon closing the
iterator, but this would require passing a `context.Context` into
`(Iterator).Close` as well as `NewIterator` which seemed potentially
controversial.

For time bounded iterators, we optionally also now return the number
of SSTables touched. This might be of interest to @danhhz.

Release note: None

tschottdorf commented:
TFTR!

bors r=petermattis

tschottdorf commented:
Acceptance build was canceled for some reason. Maybe an agent got preempted.

bors retry

tschottdorf commented:
bors r=petermattis
